vault/request_handling.go

```diff
@@ -923,6 +957,44 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
 		}
 	}
 
+	// If this is a wrapping token lookup, validate the token and produce an
+	// audit log. We perform this before c.checkToken since wrapping token
+	// validation performs a token store lookup, and can potentially modify
+	// logical.Request's token-related fields.
+	switch req.Path {
+	case "sys/wrapping/lookup":
+		valid, err := c.validateWrappingToken(ctx, req)
+		// Perform audit logging before returning if there's an issue with checking
+		// the wrapping token
+		if err != nil || !valid {
+			// We log the Auth object like so here since the wrapping token can
+			// come from the header, which gets set as the ClientToken
+			auth := &logical.Auth{
+				ClientToken: req.ClientToken,
+				Accessor:    req.ClientTokenAccessor,
+			}
+
+			logInput := &audit.LogInput{
+				Auth:               auth,
+				Request:            req,
+				NonHMACReqDataKeys: nonHMACReqDataKeys,
+			}
+			if err := c.auditBroker.LogRequest(ctx, logInput, c.auditedHeaders); err != nil {
+				c.logger.Error("failed to audit request", "path", req.Path, "error", err)
+			}
+
+			switch {
+			case err != nil:
+				return nil, auth, errwrap.Wrapf("error validating wrapping token: {{err}}", err)
+			case !valid:
+				return logical.ErrorResponse(consts.ErrInvalidWrappingToken.Error()), auth, logical.ErrInvalidRequest
+
+			}
+		}
+	}
```

Comment:
Instead of moving the code here and duplicating it, could we instead keep the validation call where it is and add audit logging into `'ValidateWrappingToken`? Since it's a core function it has access to the audit broker. WDYT?

Reply:
I am not sure if `ValidateWrappingToken` is considered a hot path. The clean way to do it in that func is with defer since there are multiple places were we return either an error or false, but defer can be relatively slow.

Reply:
Addressed in 37556fb.